Modularise examples test_suite.sh

benedikt-voelkel opened the pull request:
The examples/test_suite.sh now gives a better overview of passed/failed
tests including the concept of multiple "test cases" for each example. A
test (with multiple test cases) can be added via

start_test "SOME TEST MESSAGE"
run_test_case "the execution string"
run_test_case "another execution string"
...
run_test_case "the last execution string for this test"
finish_test "the_log_file.out"

Furthermore, the examples to be tested can be explicitly chosen using
the new command line argument --examples. For instance the examples E01
and E03 are tested with

$> examples/test_suite.sh --examples="E01 E03"

ihrivnac commented:
Thank you for this proposal. I like most of changes proposed, but not all as they are proposed.

- Added options --examples, --builddir, --garfield=* are ok
- Added  functions  in test_suite.sh are ok
- Moving all outputs in the "log" directory is a good idea, on the other hand, we already keep the reference outputs from test_suite.sh in log_ref. So I would like to keep the log and log_ref symmetric, and put the output from other tests, not saved in git, in another directory, "logs", with subdirectories:
run, run_exe, text_exe and physics_lists. 
- I don't want to extend the default output of test suite with so many details, but the current output can be improved as follows (with the coloured text for Passed, Failed):
```
... Example E01
... Running test with G4, geometry via TGeo, Native navigation    ... Passed
... Running test with G4, geometry via TGeo, TGeo navigation      ... Failed
... Running test with G4, geometry via VMC,  Native navigation    ... Failed
... Running test with G4, geometry via VMC,  TGeo navigation      ... Failed
... Failed

... 1 tests passed successfully.
... 3 tests failed.
``` 
- Your long output can be added as optional, via a new "--debug" option.

benedikt-voelkel commented:
The examples/test_suite.sh now gives a better overview of passed/failed
tests including the concept of multiple "test cases" for each example.
Tests are now directly added inside an example directory so they can be
added and removed on the fly and are automatically detected by the
test_suite.sh script.

To add tests to "SomeNewExample" add a script

directory examples/SomeNewExample/test.sh

which looks like

start_test "SOME TEST MESSAGE"
run_test_case "the execution string"
run_test_case "another execution string"
...
run_test_case "the last execution string for this test"
finish_test "the_log_file.out"

New command line args of test_suite.sh:
1) --examples
    Choose examples to be tested, for instance, the examples E01 and E03
    are tested with
    $> examples/test_suite.sh --examples="E01 E03"
2) --debug
    Turn on additional verbosity
3) --cmp-ref
    Print recursive diff of directories log/ and /log_ref

The examples/run_suite.sh now handles E03a and E03b correctly as well.

benedikt-voelkel commented:
@ihrivnac 
The further modularisation of the tests is a proposal. What do you think? If you now want to add a new test for Monopole you just have to add a test.sh in that directory and that's enough. Maybe you can give it a shot? I have the fall back which is basically the same I had in the former PR, however changed according to your former requests.

log/ and log_ref/ are now symmetric. run_suite.sh writes to logs_external/run_suite

ihrivnac commented:
HI @benedikt-voelkel ,

Thank you for the update. Though I am usually in favor of modularisation of the software, in this case, I would like to stay with the simpler version of the script defined in a single file. As the purpose of the examples is to serve both as an example for users and a test, I don't want to add the scripts, which can be run only from outside, and so which are not of the interest of the user. 

The examples are not added so often so adding a few lines of commands in the test suites is minimal effort wrt the effort to its developing & documenting. 

Could you, please, go back to your version based on your original PR and modified just for my suggestion in this thread?

Thank you,

ihrivnac commented:
Hi @benedikt-voelkel,

Thank you for the update. Could you, please, remove the updates for run_suite.sh completely, as I will simplify it in a bit different way. test_suite.sh is now fine; I will have a few minor fixes, but these can be applied after your version is merged.

Thank you,

benedikt-voelkel commented:
Hej @ihrivnac ,

I just removed the changes from run_suite.sh
